ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1239 @@
+
+                    update_operands(inputs,
+                                    input_subscripts,
+                                    input_ind1,
+                                    input_ind2,
+                                    contract_output,
+                                    resultant_subscript);
+                }
+
+                template <typename T>
+                void einsum_impl(const HostTensorVector& inputs,
+                                 const HostTensorVector& outputs,
+                                 const std::string& equation)
+                {
+                    std::vector<std::string> input_subscripts;
+                    std::string output_subscript;
+                    ngraph::opset7::Einsum::parse_equation(
+                        equation, input_subscripts, output_subscript);
+
+                    // compute einsum path that is used to contract a pair of operands
+                    // in more optimal order
+                    size_t num_inputs = inputs.size();
+                    auto einsum_path = compute_einsum_path(num_inputs);
+
+                    HostTensorVector int_inputs = inputs;
+
+                    // contract inputs by Einsum until just one is remained
+                    for (auto const& inds_pair : einsum_path)
+                    {
+                        contract_two_inputs<T>(int_inputs,
+                                               input_subscripts,
+                                               output_subscript,
+                                               inds_pair.first,
+                                               inds_pair.second);
+                    }
+
+                    NGRAPH_CHECK(int_inputs.size() == 1);
+
+                    // extract diagonal for the single operand
+                    extract_diagonal<T>(int_inputs, input_subscripts, 0);
+
+                    // reduce dimensions for the remained input node
+                    reduce_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    // transpose dimensions to layout required by the output subscript
+                    transpose_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    auto output_shape = int_inputs[0]->get_shape();
+                    const auto& element_type = int_inputs[0]->get_element_type();
+                    const auto& buf_size = shape_size(output_shape) * element_type.size();
+                    outputs[0]->write(int_inputs[0]->get_data_ptr(), buf_size);
+                }
+
+            } // namespace einsum_details
+
+            void einsum(const HostTensorVector& outputs,
+                        const HostTensorVector& inputs,
+                        const std::string& equation)
+            {
+                NGRAPH_CHECK(inputs.size() > 0, "Einsum must accept at least one input.");
+                auto input_type = inputs[0]->get_element_type();
```

Comment:
Do we sure that all inputs have the same element type?

Reply:
I will check them

Reply:
Fixed.